pylint/checkers/classes/class_checker.py

```diff
@@ -1485,6 +1536,24 @@ def _check_functools_or_not(self, decorator: nodes.Attribute) -> bool:
 
         return "functools" in dict(import_node.names)
 
+    def _has_valid_slots(self, node: nodes.ClassDef) -> bool:
```

Comment:
I'm on mobile but this seems very similar to `_check_slots`. I'm a bit worried that we're duplicating checks and run the risk of code drift.

Have you considered refactoring the other method to serve both purposes?

Reply:
I agree that `_check_slots` and `_has_valid_slots` are similar and that it would be best not to duplicate these checks. Refactoring is complicated by the fact that `_check_slots` is doing a few jobs at the same time:

- checking whether `__slots__` is valid
- if not, add 1 of 2 possible messages based on the reason why it isn't valid
- If `__slots__` looks valid, get the slot items and apply 2 further checks.

The difference between `_check_slots` and `_has_valid_slots` are that `_check_slots` cares about the reason why `__slots__` isn't valid, whereas `_has_valid_slots` only detects whether `__slots__` is valid. It might be wise to introduce a method that serves both purposes e.g. `_validate_slots` that returns an `Enum` that either reports a valid `__slots__` or says why it is invalid. This would allow you to separate the multiple concerns of the `_check_slots` method a little more clearly.

The question is whether refactoring `_check_slots` is within the scope of this MR. To me that feels like quite a big refactor that should be its own MR. I would propose that if the present MR is accepted an issue should then be opened to highlight the need to refactor these slot checks to reduce the duplication. I'd be happy to work on that issue as a separate MR.